File: code/286/solution.go
package main /*{{{*/

import "fmt"

var dp = [51][51]float64{}

func cal(a float64) float64 {
	for i := range dp {
		for j := range dp[i] {
			dp[i][j] = 0
		}
	}
	dp[0][0] = 1
	for i := 1; i <= 50; i++ {
		for j := 0; j < 50; j++ {
			dp[i][j] += dp[i-1][j] * float64(i) / a
			if j > 0 {
				dp[i][j] += dp[i-1][j-1] * (1 - float64(i)/a)
			}
		}
	}
	return dp[50][20]
}

func main() {
	l, r := float64(50), float64(100)
	for r-l >= 1e-12 {
		if cal((l+r)/2) > 0.02 {
			l = (l + r) / 2
		} else {
			r = (l + r) / 2
		}
	}
	fmt.Printf("%.10f\n", l)
}

File: code/286/solution.py
# -*- coding: UTF-8 -*-
from random import random
from math import *
from decimal import Decimal, getcontext

getcontext().prec = 100

l = Decimal(50)
r = Decimal(100)


def cal(a):
    # dp[i][j] 表示前 i 个球投中 j 个的概率
    dp = [[Decimal(0) for i in range(51)] for j in range(51)]
    dp[0][0] = Decimal(1)
    for i in range(1, 51):
        for j in range(0, 50):
            dp[i][j] += dp[i - 1][j] * Decimal(i) / a
            if j > 0:
                dp[i][j] += dp[i - 1][j - 1] * (Decimal(1) - Decimal(i) / a)
    return dp[50][20]


while r - l >= 1e-15:
    mid = (l + r) / Decimal(2)
    if cal(mid) > 0.02:
        l = mid
    else:
        r = mid
print(l.quantize(Decimal("0.0000000001")))
